bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+			 multiline ? " " : "{");
+	size -= strlen(buf);
+	buf += strlen(buf);
+
+	if (CHECK_FLAG(node_descr->flags,
+		       BGP_NLRI_TLV_NODE_DESCR_AUTONOMOUS_SYSTEM)) {
+		snprintf(buf, size, "%sAS:%u", first ? "" : " ",
+			 node_descr->autonomous_system);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(node_descr->flags, BGP_NLRI_TLV_NODE_DESCR_BGP_LS_ID)) {
+		snprintf(buf, size, "%sID:%u", first ? "" : " ",
+			 node_descr->bgp_ls_id);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(node_descr->flags, BGP_NLRI_TLV_NODE_DESCR_AREA_ID)) {
+		snprintf(buf, size, "%sArea:%u", first ? "" : " ",
+			 node_descr->area_id);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(node_descr->flags,
+		       BGP_NLRI_TLV_NODE_DESCR_IGP_ROUTER_ID)) {
+		switch (node_descr->igp_router_id_size) {
+		case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_NON_PSEUDOWIRE_SIZE:
+			snprintfrr(buf, size, "%sRtr:%pSY", first ? "" : " ",
+				   (uint8_t *)&node_descr->igp_router_id);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_PSEUDOWIRE_SIZE:
+			snprintfrr(buf, size, "%sRtr:%pPN", first ? "" : " ",
+				   (uint8_t *)&node_descr->igp_router_id);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_NON_PSEUDOWIRE_SIZE:
+			snprintfrr(buf, size, "%sRtr:%pI4", first ? "" : " ",
+				   (in_addr_t *)&node_descr->igp_router_id);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_PSEUDOWIRE_SIZE:
+			snprintfrr(
+				buf, size, "%sRtr:%pI4:%pI4", first ? "" : " ",
+				(in_addr_t *)&node_descr->igp_router_id,
+				((in_addr_t *)&node_descr->igp_router_id + 1));
+			break;
+		}
+
+		size -= strlen(buf);
+		buf += strlen(buf);
+	}
+
+	if (!multiline) {
+		snprintf(buf, size, "}");
+		size -= strlen(buf);
+	}
+
+	return size;
+}
+
+static uint16_t
+bgp_linkstate_nlri_node_prefixlen(struct bgp_linkstate_type_node *p_node)
+{
+	return sizeof(struct bgp_linkstate_type_node);
+}
+
+static int bgp_linkstate_nlri_node_decode(struct bgp_linkstate_type_node **pp,
+					  uint8_t *pnt, uint16_t len)
+{
+	int ret = BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+	struct bgp_linkstate_type_node *p_node = *pp;
+	uint8_t *lim = pnt + len;
+	uint16_t type, length;
+
+	p_node->proto = pnt_decode8(&pnt);
+	p_node->identifier = pnt_decode64(&pnt);
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+		switch (type) {
+		case BGP_LS_TLV_LOCAL_NODE_DESCRIPTORS:
+			if (p_node->local_node_descr.flags != 0)
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(p_node->local_node_descr.flags,
+				 BGP_NLRI_TLV_NODE_DESCR_LOCAL_NODE);
+			ret = bgp_linkstate_nlri_node_descriptor_decode(
+				&p_node->local_node_descr, pnt, length);
+			break;
+		default:
+			zlog_err("%s: received an unexpected TLV type %hu",
+				 __func__, type);
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+		}
```

Comment:
dont return error if unknown tlv.

    record the TLV type, and return NLRI_PARSE_OK;
    use conditional debugging to output the message.

Reply:
see https://github.com/FRRouting/frr/pull/12649#discussion_r1169925019